receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go

```diff
@@ -0,0 +1,352 @@
+		if err != nil {
+			return nil, fmt.Errorf("error creating process include filters: %v", err)
+		}
+	}
+
+	if len(cfg.Exclude.Names) > 0 {
+		scraper.excludeFS, err = filterset.CreateFilterSet(cfg.Exclude.Names, &cfg.Exclude.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process exclude filters: %v", err)
+		}
+	}
+
+	return scraper, nil
+}
+
+// Initialize
+func (s *scraper) Initialize(_ context.Context) error {
+	bootTime, err := host.BootTime()
+	if err != nil {
+		return err
+	}
+
+	s.startTime = pdata.TimestampUnixNano(bootTime)
+	return nil
+}
+
+// Close
+func (s *scraper) Close(_ context.Context) error {
+	return nil
+}
+
+// ScrapeMetrics
+func (s *scraper) ScrapeMetrics(ctx context.Context) (pdata.MetricSlice, error) {
+	_, span := trace.StartSpan(ctx, "processscraper.ScrapeMetrics")
+	defer span.End()
+
+	var errors []error
+	metrics := pdata.NewMetricSlice()
+
+	processes, err := s.getProcesses()
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	if err = scrapeAndAppendCPUUsageMetric(metrics, s.startTime, processes); err != nil {
+		errors = append(errors, err)
+	}
+
+	if err = scrapeAndAppendMemoryUsageMetric(metrics, s.startTime, processes); err != nil {
+		errors = append(errors, err)
+	}
+
+	if err = scrapeAndAppendDiskBytesMetric(metrics, s.startTime, processes); err != nil {
+		errors = append(errors, err)
+	}
+
+	if len(errors) > 0 {
+		return metrics, componenterror.CombineErrors(errors)
+	}
+
+	return metrics, nil
+}
+
+// getProcesses returns a slice of processMetadata, including handles,
+// for all currently running processes. If errors occur obtaining
+// information for some processes, an error will be returned, but any
+// processes that were successfully obtained will still be returned.
+func (s *scraper) getProcesses() ([]*processMetadata, error) {
+	handles, err := s.getProcessHandles()
+	if err != nil {
+		return nil, err
+	}
+
+	var errors []error
+	metadata := make([]*processMetadata, 0, handles.Len())
+	for i := 0; i < handles.Len(); i++ {
+		pid := handles.Pid(i)
+		handle := handles.At(i)
+		name, err := getProcessName(handle)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process name for pid %v: %v", pid, err))
+			continue
+		}
+
+		// filter processes by name
+		if (s.includeFS != nil && !s.includeFS.Matches(name)) ||
+			(s.excludeFS != nil && s.excludeFS.Matches(name)) {
+			continue
+		}
+
+		md := &processMetadata{}
+		err = initializeProcessMetadata(md, pid, name, handle)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		metadata = append(metadata, md)
+	}
+
+	if len(errors) > 0 {
+		return metadata, componenterror.CombineErrors(errors)
+	}
+
+	return metadata, nil
+}
+
+func getProcessName(proc processHandle) (string, error) {
+	if runtime.GOOS != "windows" {
+		return proc.Name()
+	}
+
+	// calling proc.Name() is currently prohibitively expensive on Windows, so use the exe name instead (mirrors psutil)
```

Comment:
Is it because it's a WMI call or other reasons?

Reply:
No, in this case it calls https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot (using the TH32CS_SNAPPROCESS flag). I don't know why this is so slow but from my testing it was taking around 1s (on a machine running lots of processes)

Reply:
Could we maybe call this instead? https://docs.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getmodulebasenamew

Reply:
Ah, yeah CreateToolhelp32Snapshot sounds rather expensive. What you have seems fine to me.